### PR TITLE
Fix docblocks and imports in FormController behaviour.

### PR DESCRIPTION
**Pull request by jacobdekeizer**
Add correct parameter, return type and exception docblocks for FormController behaviour.
Import whole namespace for facade so we can follow the code in the editor.

### CODE REVIEW

File: modules/backend/behaviors/FormController.php [hunk 1]
```diff
@@ -1,17 +1,16 @@
...
+use Backend\Facades\Backend;
 use Exception;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Event;
+use Illuminate\Support\Facades\Lang;
+use Illuminate\Support\Facades\Redirect;
+use October\Rain\Exception\ApplicationException;
+use October\Rain\Router\Helper as RouterHelper;
+use October\Rain\Support\Facades\Flash;
+use October\Rain\Support\Str;
```
Comment (LukeTowers) [inline modules/backend/behaviors/FormController.php]: We will not be accepting these namespace changes, if you want to be able to click through to the source classes in your IDE then use https://octobercms.com/plugin/flynsarmy-idehelper. Fully specifying the namespaces to facades completely defeats the point and makes the application more brittle. An excellent example of this is the fact that Laravel removed the Input facade and we added it back, but any code that was directly referencing the Laravel version is now going to be broken in the L6 upgrade. 

@bennothommo in relation to that, do you think we can setup a class alias for the old facade to improve compatibility?

**Comment by jacobdekeizer** [comment]
@LukeTowers Updated my pr

File: modules/backend/behaviors/FormController.php [hunk 2]
```diff
@@ -763,8 +771,10 @@ public function formFindModelObject($recordId)
         /*
          * Prepare query and find model record
          */
+        /** @var \October\Rain\Database\Builder $query */
         $query = $model->newQuery();
         $this->controller->formExtendQuery($query);
+        /** @var \October\Rain\Database\Model $result */
```
Comment (LukeTowers) [inline modules/backend/behaviors/FormController.php]: What is the point of these comments?

File: modules/backend/behaviors/FormController.php [hunk 3]
```diff
@@ -8,7 +8,6 @@
 use Redirect;
 use Backend;
 use Backend\Classes\ControllerBehavior;
-use October\Rain\Html\Helper as HtmlHelper;
```
Comment (LukeTowers) [inline modules/backend/behaviors/FormController.php]: Why did you remove this?

Comment (jacobdekeizer) [reply modules/backend/behaviors/FormController.php]: Because it is unused

File: modules/backend/behaviors/FormController.php (hunk 2 above)
Comment (jacobdekeizer) [reply modules/backend/behaviors/FormController.php]: Because it is an illuminate function which returns as type hint an illuminate model instead of the correct october model

Comment (jacobdekeizer) [reply modules/backend/behaviors/FormController.php]: With this docblock the typehint is restored to the correct model

**Comment by bennothommo** [comment]
@jacobdekeizer I think the reasoning for the `Model` typehint was because the `FormController` also supports Halcyon models (eg. `October\Rain\Halcyon\Model`), so I think the model typehints would need to reflect that those methods and properties can take both.

If you want a real challenge, you could propose an interface/contract for the Rain library that covers both and use that as the type-hint.

**Comment by jacobdekeizer** [comment]
@bennothommo I think it's best to add both Models as typehint for now.

**Comment by jacobdekeizer** [comment]
@bennothommo I've added the Halcyon model as typehint

File: modules/backend/behaviors/FormController.php [hunk 4]
```diff
@@ -339,7 +339,9 @@ public function update_onSave($recordId = null, $context = null)
      * `formAfterDelete`.
      *
      * @param int $recordId Record identifier
-     * @return mixed
+     * @return \Illuminate\Http\RedirectResponse|void
+     * @throws \October\Rain\Exception\ApplicationException
+     * @throws Exception
```
Comment (bennothommo) [inline modules/backend/behaviors/FormController.php]: @jacobdekeizer I'm leaning towards these being redundant on all methods in this class. They would make sense if we have multiple exception classes for individual pieces of the system, but we don't - we generally use `ApplicationException` and `SystemException` and that's it, so it can be a general assumption that a method *may* throw one of those exceptions at any given point.

Would you have any problems with removing them from your changes?

**Comment by bennothommo** [comment]
@jacobdekeizer Just out of curiousity, are these doc hints being generated by an automated system, or are you manually entering them in?

File: modules/backend/behaviors/FormController.php (hunk 4 above)
Comment (jacobdekeizer) [reply modules/backend/behaviors/FormController.php]: Why should we make an assumption when we can typehint that this method throws these exceptions. Phpstorm will warn you when you don't catch an exception, if we remove this phpstorm will not warn you for uncaught exceptions. So I'm against removing them, because it helps the developer with what he can expect from these methods.

**Comment by jacobdekeizer** [comment]
@bennothommo Phpstorm can autogenerate these docblocks. In my opinion all octobercms maintainers should use phpstorm, because it finds many typehint errors and other coding errors. I could make another x amount of pull requests to fix a lot of files, but I don't have the time to do that. If the maintainers use phpstorm they could see all these typehint problems themselves in advance, instead of someone else fixing it with a pull request. You could also look into tools like phpstan and mess dectector for better code.

File: modules/backend/behaviors/FormController.php (hunk 4 above)
Comment (bennothommo) [reply modules/backend/behaviors/FormController.php]: @jacobdekeizer I recognise that PHPStorm can analyse this and give hints, but on the outset, if I'm reading that, those two lines in particular tell me absolutely nothing. It doesn't give me any indication as to *why* those exceptions might be thrown, especially when you view the code itself in which you find that the exception wouldn't be thrown by any of the method's code, and that PHPStorm found that the exception was instead potentially thrown by a method 3 or 4 levels down in the call stack.

It's also presumptuous in some cases, as there's potentially many more types of exceptions that can be thrown - especially when you bring plugins into the mix.

**Comment by bennothommo** [comment]
@jacobdekeizer I appreciate the work that you have put into this so far. Perhaps for now, if its your intention to clean up a few more classes, maybe stick with the obvious cases of type hints that are incorrect - it's my hope that we can use actual parameter type hints and return types in the future, which will make most of the docblock changes redundant.

File: modules/backend/behaviors/FormController.php (hunk 4 above)
Comment (jacobdekeizer) [reply modules/backend/behaviors/FormController.php]: That's why we have to typehint all exceptions, so we (as developers) aren't surprised with undocumented exceptions.

Comment (LukeTowers) [reply modules/backend/behaviors/FormController.php]: @jacobdekeizer the point @bennothommo is trying to make is that at some point it's not possible to typehint all possible exceptions that could be thrown, especially when you start throwing event handling into the mix.

Comment (bennothommo) [reply modules/backend/behaviors/FormController.php]: @jacobdekeizer If you add context as to why those exceptions might be thrown (as per [@throws](https://docs.phpdoc.org/latest/references/phpdoc/tags/throws.html) documentation), it might make it more credible to keep these lines.

**Comment by jacobdekeizer** [comment]
@bennothommo I've added descriptions to the `@throws` exception docblocks.

File: modules/backend/behaviors/FormController.php (hunk 2 above)
Comment (bennothommo) [reply modules/backend/behaviors/FormController.php]: @jacobdekeizer I get your reasoning, but why does this need to be specified when normally you would not interact with these variables at all besides through the `formExtendQuery` and `formExtendModel` methods which are already correctly type-hinted?

Comment (jacobdekeizer) [reply modules/backend/behaviors/FormController.php]: @bennothommo You're right, these docblocks can be removed.

**Comment by bennothommo** [comment]
@jacobdekeizer I'm happy to merge this, however, I'll check with @daftspunk if this is going to have any conflicts with the [API Docs](https://octobercms.com/docs/api) (mainly in terms of the multiple model type hints)
